src/services/TestTypesService.ts

```diff
@@ -1,220 +1,321 @@
-                    .forEach((field: keyof TestCode) => {
-                        response[field] = testCodes[0][field];
-                    });
-
-                // Populating testTypeClassification that is found in testType, not testCode
-                if (filterExpression.fields.includes("testTypeClassification")) {
-                    response.testTypeClassification = testType.testTypeClassification;
-                }
-                return response;
-            });
-    }
+        return this.sort(filteredTestType);
+      })
+      .catch((error) => {
+        if (!(error instanceof HTTPError)) {
+          console.error(error);
+          error.statusCode = 500;
+          error.body = "Internal Server Error";
+        }
 
-    /**
-     * returns null if the test-type was not found
-     * @param id
-     * @param testTypes
-     */
-    public findTestType({ id, testTypes }: { id: string; testTypes: any; }) {
-        for (const testType of testTypes) {
-            // for (let i = 0; i < testTypes.length; i++) {
-            //   const testType: any = testTypes[i];
-            if (testType.hasOwnProperty("nextTestTypesOrCategories")) {
-                const childrenTestType: any = this.findTestType({ id, testTypes: testType.nextTestTypesOrCategories });
-
-                if (childrenTestType != null) {
-                    return childrenTestType;
-                }
-            } else if (testType.hasOwnProperty("id") && testType.id === id) {
-                return testType;
-            }
+        throw new HTTPError(error.statusCode, error.body);
+      });
+  }
+
+  public getTestTypesById(id: string, filterExpression: any) {
+    return this.testTypesDAO
+      .getAll()
+      .then((data) => {
+        if (data.Count === 0) {
+          throwResourceNotFound();
         }
-        return null;
-    }
 
-    public sort(testTypes: ITestType[] | any) {
-        // Pass by value
-        const testTypeArray = testTypes;
+        return data.Items;
+      })
+      .then((testTypes) => {
+        return this.findTestType({ id, testTypes });
+      })
+      .then((testType) => {
+        if (testType === null) {
+          throwResourceNotFound();
+        }
+        let testCodes: TestCode[] = testType.testCodes;
+
+        testCodes = testCodes
+          .filter((testCode) => {
+            // filter by vehicleType if present in DB, otherwise skip
+            return testCode.forVehicleType
+              ? this.fieldInFilterExpressionMatchesTheOneInTestCode(
+                  testCode,
+                  filterExpression,
+                  "forVehicleType"
+                )
+              : true;
+          })
+          .filter((testCode) => {
+            // filter by vehicleSize if present in DB & in request, otherwise skip
+            return testCode.forVehicleSize && filterExpression.vehicleSize
```

Comment:
Had a thought about it, we might be able to refactor that by currying testCode in the filter operation since we are returning truthies values that could be coerced if my understanding is correct.

```
// Map methods return callbacks so we can use a closure to carry the item to be iterated over
// we can prepare multiple util functions that will be used as operations in the pipe

const byTestCodeByVehicleSize = 
  (filterExpression, param) =>
    (testCode) => 
       this.fieldsInFilterExpressionMatchesTheOneInTestCode(testCode, filterExpression, param)
```

```
// in the filter pipe
testCodes
  .filter(byTestCodeByVehicleSize(filterExpression, "forVehicleSize")) // returns a callback that we curried
  .filter(...) // we carry on chaining the filter operations
```

Reply:
as discussed, resolving as this was part of existing code and showing during auto formatting https://jira.dvsacloud.uk/browse/CVSB-17166

Reply:
as discussed, resolving as this was part of existing code and showing during auto formatting https://jira.dvsacloud.uk/browse/CVSB-17166